modules/fluentd-forwarder/apis/composition.yaml

```diff
@@ -0,0 +1,168 @@
+                    configMaps:
+                      useDefaults:
+                        containersInputConf: false
+                        forwardInputConf: false
+                        monitoringConf: false
+                        outputConf: false
+                        systemConf: false
+                        systemInputConf: true
+                    # Liveliness probe reverted from upstream changes to prevent that it fails if just one buffer is inactive.
+                    # E.g. the authlog buffer will often be inactive, so it would make the probe fail.
+                    livenessProbe:
+                      kind:
+                        exec:
+                          command:
+                          # Liveness probe is aimed to help in situations where fluentd
+                          # silently hangs for no apparent reasons until manual restart.
+                          # The idea of this probe is that if fluentd is not queueing or
+                          # flushing chunks for 5 minutes, something is not right. If
+                          # you want to change the fluentd configuration, reducing amount of
+                          # logs fluentd collects, consider changing the threshold or turning
+                          # liveness probe off completely.
+                          - '/bin/sh'
+                          - '-c'
+                          - >
+                            LIVENESS_THRESHOLD_SECONDS=${LIVENESS_THRESHOLD_SECONDS:-300};
+                            STUCK_THRESHOLD_SECONDS=${STUCK_THRESHOLD_SECONDS:-900};
+                            if [ ! -e /var/log/fluentd-buffers ];
+                            then
+                              echo "Expected directory /var/log/fluentd-buffers does not exist. This is likely a configuration issue.";
+                              exit 1;
+                            fi;
+                            touch -d "${STUCK_THRESHOLD_SECONDS} seconds ago" /tmp/marker-stuck;
+                            if [ -z "$(find /var/log/fluentd-buffers -type d -newer /tmp/marker-stuck -print -quit)" ];
+                            then
+                              echo "Elasticsearch buffers found stuck longer than $STUCK_THRESHOLD_SECONDS seconds. Clearing buffers."
+                              rm -rf /var/log/fluentd-buffers;
+                              exit 1;
+                            fi;
+                            touch -d "${LIVENESS_THRESHOLD_SECONDS} seconds ago" /tmp/marker-liveness;
+                            if [ -z "$(find /var/log/fluentd-buffers -type d -newer /tmp/marker-liveness -print -quit)" ];
+                            then
+                              echo "Elasticsearch buffers found stuck longer than $LIVENESS_THRESHOLD_SECONDS seconds."
+                              exit 1;
+                            fi;
+            patches:
+              # log-write capability
+              - type: FromEnvironmentFieldPath
```

Comment:
More for my understanding: Why did you use `FromEnvironmentFieldPath` instead of referencing a Connection Detail? Is this due to a Crossplane limitation or a deliberate design choice?

Reply:
Good and very observant question!

You can't patch directly from connection details but you can achieve it by referencing a secret and patch the name of the secret. See this example: https://github.com/crossplane-contrib/provider-kubernetes/blob/e6f347a8a1deeda4703805f88e1305f605cf2b56/examples/in-composition/composition.yaml#L118-L137

The same could be done here but for the PoC I wanted to create a "capability interface" which in this case is the EnvironmentConfig.
* The producer (OpenSearch) writes to the interface.
* The receiver (FluentdForwarder) reads from the interface.

It can be modeled in many different ways. This was just one I picked.